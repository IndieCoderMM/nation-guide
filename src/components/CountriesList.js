import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { BsFillMicFill } from 'react-icons/bs';
import { AiFillSetting } from 'react-icons/ai';
import { getAllCountries } from '../redux/countries';
import CountryCard from './CountryCard';
import SortingBox from './SortingBox';
import SearchBox from './SearchBox';
import DisplayStatus from './DisplayStatus';

const CountriesList = () => {
  const data = useSelector((state) => state.data);
  const status = useSelector((state) => state.status);
  const [count, setCount] = useState(11);
  const [query, setQuery] = useState('');
  const [sorter, setSorter] = useState('area-d');
  const dispatch = useDispatch();

  useEffect(() => {
    if (status === 'idle') dispatch(getAllCountries());
  }, [status, dispatch]);

  // Search Feature
  const countries = query.trim()
    ? [...data].filter((c) => c.name.common.toLowerCase().includes(query))
    : [...data];

  // Sort Feature
  if (sorter === 'area-d') {
    countries.sort((a, b) => b.area - a.area);
  } else if (sorter === 'area-a') {
    countries.sort((a, b) => a.area - b.area);
  } else if (sorter === 'name-d') {
    countries.sort((a, b) => (a.name.common > b.name.common ? 1 : -1));
  } else if (sorter === 'name-a') {
    countries.sort((a, b) => (a.name.common < b.name.common ? 1 : -1));
  }

  return (
    <section className="homepage">
      <div className="toolbar">
        <h1 className="brand">
          <Link to="/">NationGuide</Link>
        </h1>
        <SearchBox query={query} setQuery={setQuery} />
        <div className="d-row">
          <BsFillMicFill />
          <AiFillSetting />
        </div>
      </div>
      {countries.length > 0 && (
        <CountryCard
          className="feature-item"
          name={countries[0].name}
          area={countries[0].area}
          flagPng={countries[0].flags.png}
          flagAlt={countries[0].flags.alt}
        />
      )}
      <SortingBox sorter={sorter} setSorter={setSorter} />
      <DisplayStatus status={status} />
      {countries.length === 0 && <DisplayStatus status="notfound" />}
      <div className="main-grid">
        {countries.slice(1, count).map((c) => (
          <CountryCard
            key={c.name.common}
            className="grid-item"
            name={c.name}
            area={c.area}
            flagPng={c.flags.png}
            flapAlt={c.flags.alt}
          />
        ))}
      </div>
      {count < countries.length && (
        <div className="btn-container">
          <button
            className="more-btn"
            type="button"
            onClick={() => setCount((count) => count + 10)}
          >
            Load More
          </button>
        </div>
      )}
    </section>
  );
};

export default CountriesList;
