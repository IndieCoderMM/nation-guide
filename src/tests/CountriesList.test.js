import React from 'react';
import axios from 'axios';
import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import CountriesList from '../components/CountriesList';

jest.mock('axios');

const responseData = [
  {
    name: {
      common: 'Canada',
      nativeName: {
        eng: {
          official: 'Canada',
          common: 'Canada',
        },
      },
    },
    currencies: {
      CAD: {
        name: 'Canadian dollar',
        symbol: '$',
      },
    },
    capital: ['Ottawa'],
    altSpellings: ['CA'],
    region: 'Americas',
    latlng: [60.0, -95.0],
    area: 9984670.0,
    flag: 'ðŸ‡¨ðŸ‡¦',
    population: 38005238,

    timezones: ['UTC-08:00', 'UTC-07:00'],
    continents: ['North America'],
    flags: {
      png: 'https://flagcdn.com/w320/ca.png',
      svg: 'https://flagcdn.com/ca.svg',
      alt: 'The flag of Canada is composed of a red vertical band on the hoist and fly sides and a central white square that is twice the width of the vertical bands. A large eleven-pointed red maple leaf is centered in the white square.',
    },
    coatOfArms: {
      png: 'https://mainfacts.com/media/images/coats_of_arms/ca.png',
      svg: 'https://mainfacts.com/media/images/coats_of_arms/ca.svg',
    },
  },
];

describe('Component for Homepage', () => {
  it('should fetch data from API on load', async () => {
    // Arrange
    axios.get.mockResolvedValue({ data: responseData });
    let asFragment;
    // Act
    await act(() => {
      const component = render(
        <BrowserRouter>
          <Provider store={store}>
            <CountriesList />
          </Provider>
          ,
        </BrowserRouter>,
      );
      asFragment = component.asFragment;
    });
    await screen.findByText(responseData[0].name.common);
    // Assert
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(responseData[0].name.common)).toBeInTheDocument();
  });
});
