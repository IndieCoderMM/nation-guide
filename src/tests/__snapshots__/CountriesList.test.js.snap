// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component for Homepage should fetch data from API on load 1`] = `
<DocumentFragment>
  <section
    class="homepage"
  >
    <div
      class="toolbar"
    >
      <h1
        class="brand"
      >
        <a
          href="/"
        >
          NationGuide
        </a>
      </h1>
      <form
        class="search-box"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          class="search-input"
          maxlength="50"
          placeholder="Search country by name"
          type="text"
          value=""
        />
      </form>
      <div
        class="d-row"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 3a3 3 0 0 1 6 0v5a3 3 0 0 1-6 0V3z"
          />
          <path
            d="M3.5 6.5A.5.5 0 0 1 4 7v1a4 4 0 0 0 8 0V7a.5.5 0 0 1 1 0v1a5 5 0 0 1-4.5 4.975V15h3a.5.5 0 0 1 0 1h-7a.5.5 0 0 1 0-1h3v-2.025A5 5 0 0 1 3 8V7a.5.5 0 0 1 .5-.5z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512.5 390.6c-29.9 0-57.9 11.6-79.1 32.8-21.1 21.2-32.8 49.2-32.8 79.1 0 29.9 11.7 57.9 32.8 79.1 21.2 21.1 49.2 32.8 79.1 32.8 29.9 0 57.9-11.7 79.1-32.8 21.1-21.2 32.8-49.2 32.8-79.1 0-29.9-11.7-57.9-32.8-79.1a110.96 110.96 0 0 0-79.1-32.8zm412.3 235.5l-65.4-55.9c3.1-19 4.7-38.4 4.7-57.7s-1.6-38.8-4.7-57.7l65.4-55.9a32.03 32.03 0 0 0 9.3-35.2l-.9-2.6a442.5 442.5 0 0 0-79.6-137.7l-1.8-2.1a32.12 32.12 0 0 0-35.1-9.5l-81.2 28.9c-30-24.6-63.4-44-99.6-57.5l-15.7-84.9a32.05 32.05 0 0 0-25.8-25.7l-2.7-.5c-52-9.4-106.8-9.4-158.8 0l-2.7.5a32.05 32.05 0 0 0-25.8 25.7l-15.8 85.3a353.44 353.44 0 0 0-98.9 57.3l-81.8-29.1a32 32 0 0 0-35.1 9.5l-1.8 2.1a445.93 445.93 0 0 0-79.6 137.7l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.2 56.5c-3.1 18.8-4.6 38-4.6 57 0 19.2 1.5 38.4 4.6 57l-66 56.5a32.03 32.03 0 0 0-9.3 35.2l.9 2.6c18.1 50.3 44.8 96.8 79.6 137.7l1.8 2.1a32.12 32.12 0 0 0 35.1 9.5l81.8-29.1c29.8 24.5 63 43.9 98.9 57.3l15.8 85.3a32.05 32.05 0 0 0 25.8 25.7l2.7.5a448.27 448.27 0 0 0 158.8 0l2.7-.5a32.05 32.05 0 0 0 25.8-25.7l15.7-84.9c36.2-13.6 69.6-32.9 99.6-57.5l81.2 28.9a32 32 0 0 0 35.1-9.5l1.8-2.1c34.8-41.1 61.5-87.4 79.6-137.7l.9-2.6c4.3-12.4.6-26.3-9.5-35zm-412.3 52.2c-97.1 0-175.8-78.7-175.8-175.8s78.7-175.8 175.8-175.8 175.8 78.7 175.8 175.8-78.7 175.8-175.8 175.8z"
          />
        </svg>
      </div>
    </div>
    <a
      class="feature-item"
      href="/country/Canada"
    >
      <div
        class="frame"
      >
        <img
          alt="The flag of Canada is composed of a red vertical band on the hoist and fly sides and a central white square that is twice the width of the vertical bands. A large eleven-pointed red maple leaf is centered in the white square."
          src="https://flagcdn.com/w320/ca.png"
        />
      </div>
      <div>
        <h3>
          Canada
        </h3>
        <p>
          9984670 kmÂ²
        </p>
      </div>
    </a>
    <div
      class="bar-item"
    >
      <select
        class="sorting-box"
      >
        <option
          value="area-d"
        >
          Sort by Area [D]
        </option>
        <option
          value="area-a"
        >
          Sort by Area [A]
        </option>
        <option
          value="name-d"
        >
          Sort by Name [A]
        </option>
        <option
          value="name-a"
        >
          Sort by Name [D]
        </option>
      </select>
    </div>
    <div
      class="main-grid"
    />
  </section>
  ,
</DocumentFragment>
`;
